From 62ba80512653be5149b5ae535d235b2ce7ac1afb Mon Sep 17 00:00:00 2001
From: Jack Ping CHNG <jack.ping.chng@linux.intel.com>
Date: Sun, 26 Apr 2020 11:21:13 +0800
Subject: [PATCH] DRV_DPM_SW-238: DPM Enhance handling of dp_register_subif_ext
 - deregistration

Handles clearing of MAC and removal of bridge port and bridge if net device is
not bring down during deregistration of subif.
---
 drivers/net/datapath/dpm/datapath_instance.c | 31 +++++++++++++++++++++++-----
 1 file changed, 26 insertions(+), 5 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath_instance.c b/drivers/net/datapath/dpm/datapath_instance.c
index cb102322ac59..62ed4b0764b5 100644
--- a/drivers/net/datapath/dpm/datapath_instance.c
+++ b/drivers/net/datapath/dpm/datapath_instance.c
@@ -262,9 +262,6 @@ int dp_inst_add_dev(struct net_device *dev, char *subif_name, int inst,
 		dp_dev = devm_kzalloc(&g_dp_dev->dev, sizeof(*dp_dev),
 				      GFP_ATOMIC);
 		if (dp_dev) {
-			dp_dev->count = 0;
-			dp_dev->subif_name[0] = 0;
-			dp_dev->fid = 0;
 			INIT_LIST_HEAD(&dp_dev->ctp_list);
 			new_f = 1;
 		}
@@ -303,8 +300,7 @@ int dp_inst_add_dev(struct net_device *dev, char *subif_name, int inst,
 			      port_info->vap_mask);
 		sif = get_dp_port_subif(port_info, vap);
 #if IS_ENABLED(CONFIG_INTEL_DATAPATH_SWITCHDEV)
-		if ((!(flag & DP_F_SUBIF_LOGICAL)) &&
-		    (sif->swdev_en == 1))
+		if ((!(flag & DP_F_SUBIF_LOGICAL)) && sif->swdev_en == 1)
 			dp_register_switchdev_ops(dp_dev, dev);
 #endif
 #if IS_ENABLED(CONFIG_INTEL_DATAPATH_PTP1588)
@@ -355,6 +351,31 @@ int dp_inst_del_dev(struct net_device *dev, char *subif_name, int inst,
 			dp_dev->count--;
 
 			if (!dp_dev->count) {
+#if IS_ENABLED(CONFIG_INTEL_DATAPATH_SWITCHDEV)
+				if (netif_is_bridge_port(dev)) {
+					struct net_device *br_dev;
+					struct inst_info *dp_info;
+					int inst, lock = 0;
+					u8 *addr;
+
+					inst = dp_dev->inst;
+					addr = (u8 *)dev->dev_addr;
+					dp_info = get_dp_prop_info(inst);
+					dp_info->dp_mac_reset(0, dp_dev->fid,
+							      inst, addr);
+					if (!rtnl_is_locked()) {
+						rtnl_lock();
+						lock = 1;
+					}
+					br_dev =
+					      netdev_master_upper_dev_get(dev);
+					if (br_dev)
+						dp_del_br_if(dev, br_dev, inst,
+							     dp_dev->bp);
+					if (lock)
+						rtnl_unlock();
+				}
+#endif
 				hlist_del(&dp_dev->hlist);
 				if (dev)
 					dp_ops_reset(dp_dev, dev);
